File: 2024_12/leetcode2024-12-15.py
from typing import List
from collections import Counter

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        cnt = Counter(arr)
        res,tot = 0,0
        for i,c in cnt.most_common():
            res += 1
            tot += c
            if tot >= len(arr)//2:
                break
        return res

if __name__ == "__main__":
    s = Solution()
    arr = [3,5,3,5,3,3,2,7,5,2]
    print(s.minSetSize(arr))

File: leetcode2024/2024_12/leetcode2024-12-15.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> cnt;
        for (int i : arr){
            cnt[i]++;
        }
        vector<int> cnts;
        for (auto& kv : cnt){
            cnts.push_back(kv.second);
        }
        sort(cnts.begin(), cnts.end(), greater<int>());
        int res = 0, tot = 0;
        for (int i : cnts){
            res++;
            tot += i;
            if (tot >= arr.size()/2){
                break;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> arr = {3,3,5,5,5,3,3,2,7,2};
    cout << s.minSetSize(arr) << endl;
}
